src/utils/server-utils.ts

```diff
@@ -52,6 +52,46 @@ function formatCapabilitiesInfo(
   }, {});
 }
 
+type FlatLayer = {
+  Name: {
+    _text: string;
+    [key: string]: any;
+  };
+  Dimension: {
+    _text: string;
+    [key: string]: any;
+  };
+  [key: string]: any;
+};
+
+type FlatLayerContainer = { Layer: FlatLayer; [key: string]: any };
+type LayerContainer =
+  | { Layer: LayerContainer; [key: string]: any }
+  | FlatLayerContainer[]
+  | FlatLayer[];
+
+const isArrayOfFlatLayerContainers = (
+  maybeArray: LayerContainer,
+): maybeArray is FlatLayerContainer[] => {
+  return (maybeArray as FlatLayerContainer[])[0].Layer !== undefined;
+};
+
+function flattenLayers(rawLayers: LayerContainer): FlatLayer[] {
+  if ('Layer' in rawLayers) {
+    return flattenLayers(rawLayers.Layer);
+  }
+  if (!Array.isArray(rawLayers) || rawLayers.length === 0) {
+    return [];
+  }
+  if (isArrayOfFlatLayerContainers(rawLayers)) {
+    return rawLayers.reduce(
+      (acc, { Layer }) => acc.concat(Layer),
+      [] as FlatLayer[],
+    );
+  }
+  return rawLayers as FlatLayer[];
```

Comment:
I'm confused by this cast. You're basically telling Typescript to assume that `rawLayers` is a `FlatLayer[]`. But the point of `isArrayOfFlatLayerContainers` is to test whether `rawLayers` is `FlatLayer[]`. So how do you imagine this condition ever being reached? It seems like almost by definition, if this logic branch is reached, `rawLayers` is _not_ a `FlatLayer[]` (since it failed the `isArrayOfFlatLayerContainers` test?)

Reply:
If we are ever reaching this point, then we're pretty safe to assume that it is not an array of FlatLayerContainer.

Basically the first `if` takes care of the depth of `Layers` and the other two take care of the difference between the way geoserver layers are stored and how its done in Open Data Cube

Reply:
Ah yes. I think my confusion is that I was thinking `isArrayOfFlatLayerContainers` was checking for `FlatLayer[]` rather than `FlatLayerContainer[]`.

Just curious - does TS still require you to cast this last line if you wrap it in an `else` block from the `isArrayOfFlatLayerContainers`? It _seems_ like the logic of the union type shouldn't require the cast, but TS might not be smart enough to realize that...

Reply:
It does not but I wanted to make it a bit clearer :)